src/coreclr/vm/callstubgenerator.cpp

```diff
@@ -0,0 +1,1013 @@
+                                case 16:
+                                    m_pHeader->Invoke = CallJittedMethodRet4Float;
+                                    break;
+                                default:
+                                    _ASSERTE(!"Should not get here");
+                                    break;
+                            }
+                            break;
+                        case CORINFO_HFA_ELEM_DOUBLE:
+                            switch (thReturnValueType.GetSize())
+                            {
+                                case 8:
+                                    m_pHeader->Invoke = CallJittedMethodRetDouble;
+                                    break;
+                                case 16:
+                                    m_pHeader->Invoke = CallJittedMethodRet2Double;
+                                    break;
+                                case 24:
+                                    m_pHeader->Invoke = CallJittedMethodRet3Double;
+                                    break;
+                                case 32:
+                                    m_pHeader->Invoke = CallJittedMethodRet4Double;
+                                    break;
+                                default:
+                                    _ASSERTE(!"Should not get here");
+                                    break;
+                            }
+                                break;
+                        default:
+                            _ASSERTE(!"HFA types other than float and double are not supported yet");
+                            break;
+                    }
+                }
+                else
+                {
+                    switch (thReturnValueType.GetSize())
+                    {
+                        case 1:
+                        case 2:
+                        case 4:
+                        case 8:
+                            m_pHeader->Invoke = CallJittedMethodRetI8;
+                            break;
+                        case 16:
+                            m_pHeader->Invoke = CallJittedMethodRet2I8;
+                            break;
+                        default:
+                            _ASSERTE(!"The return types that are not HFA should be <= 16 bytes in size");
+                            break;
+                    }
+                }
+#else
+                _ASSERTE(!"Struct returns by value are not supported yet");
+#endif
+                break;
+            default:
+                _ASSERTE(!"Unexpected return type");
+                break;
+        }
+    }
+
+    m_pHeader->NumRoutines = m_routineIndex + 1; // Reserve one extra slot for the target method pointer
+    m_pHeader->TotalStackSize = m_totalStackSize;
+
+    // resize the structure to its actually utilized size
+    m_pHeader = (CallStubHeader*)realloc(m_pHeader, sizeof(CallStubHeader) + m_pHeader->NumRoutines * sizeof(PCODE));
```

Comment:
The persistent memory should be allocated on the loader heap, so that we do not need to worry about freeing it.

I would build it into something like CQuickBytes, and then allocate memory chunk of the right size on loader heap to persist it, (We may want to check that it was not allocated yet before allocating the memory on loader heap to reduce chance of wasting allocations.)

Reply:
Right, that makes sense, I will do that.